make collection callbacks follow normal flow

bcoca opened the pull request:
fixes missing set_options call and adhoc and stdout processing rules

fixes #59930

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
tqm

bcoca committed:
make collections whitelist follow normal flow

 fixes missing set_options call and adhoc and stdout processing rules

bcoca committed:
deal with generator addition

bcoca committed:
avoid dupes

bcoca committed:
fix typos

bcoca committed:
also avoid dupes in whitelist itself

bcoca committed:
dont mix class refs with names

cidrblock commented:
The latest change still work because the callback name is derived from the path, but the whitelisted callback name includes the collection name, therefore set_option is still not called.

```0.0s load_callbacks: 
      C.DEFAULT_CALLBACK_WHITELIST=['nmake.jetpack.network_change_log']
 0.0s load_callbacks: callback_name='network_change_log'
```

bcoca committed:
add note

bcoca committed:
revamp

bcoca committed:
fix + better debug

bcoca committed:
build list diff

bcoca committed:
skip invalid fqcn cbs

sivel commented:
Here is another report of an underlying issue that this resolves: https://groups.google.com/g/ansible-project/c/ELXVfSngWTc/m/K7L1sqboAAAJ

It appears as though we allow loading multiple `stdout` types once we've made it to the collections 2nd pass, and this PR seems to address that.

bcoca committed:
added b ack missing import

bcoca committed:
fixed to handle redirects

also updated tests with new and more accurate skip message

bcoca committed:
removed now unused import

bcoca committed:
avoid skip message for dupes

 fix callback tests for envs with cowsay installed

bcoca committed:
removed test for 2 names loading same callback

  current code doesn't warn as this as it can happen for more reasons than previouslly
  now that we are properly processing all the plugin sources in the same list.

bcoca committed:
read test but as duplicate avoidance

nitzmahone commented on lib/ansible/executor/task_queue_manager.py:
```diff
@@ -171,49 +171,62 @@ def load_callbacks(self):
-            (callback_name, _) = os.path.splitext(os.path.basename(callback_plugin._original_path))
+
+            # try to get colleciotn world name first
+            cnames = getattr(callback_plugin, '_redirected_names', [])
```
Argh, I just realized the PluginLoader attribute stuffing (update_object) has been happening on `class_only` cases, which is broken for "same name, different behavior" cases (since subsequent loads will overwrite the class vars on the existing type object, not instance vars). It works here, since `all` is `yield`ing the plugin class immediately after calling `update_object` on it, but the whole pattern is busted for `class_only`, esp for things that might consult the class var during or after `__init__`. We've needed to rethink that for awhile, and this kinda sets it in a little more concrete. :(

bcoca replied:
We do this for connection, shell, become, actions, callbacks, etc ... if this is a problem it affects most of our plugins. But this also was not introduced in this PR and is widespread, so i'm not going to hold off this patch based on it.

nitzmahone commented on lib/ansible/executor/task_queue_manager.py:
```diff
@@ -171,49 +171,62 @@ def load_callbacks(self):
+            # try to get colleciotn world name first
+            cnames = getattr(callback_plugin, '_redirected_names', [])
+            if cnames:
+                # use 'first name' and not 'real name' as that is what will match config
```
It wasn't immediately obvious to me why this needed to happen, maybe clarify the comment to something like:

```suggestion
                # store the name the plugin was loaded as, as that's what we'll need to compare to the configured callback list later
```

nitzmahone commented on lib/ansible/executor/task_queue_manager.py:
```diff
@@ -171,49 +171,62 @@ def load_callbacks(self):
-                callback_obj, plugin_load_context = callback_loader.get_with_context(callback_plugin_name)
+            try:
+                callback_obj = callback_plugin()
                 if callback_obj:
```
This implies that constructing some callbacks doesn't result in an instance, but also doesn't raise an Exception? That doesn't sound right...

bcoca replied:
IIRC this mostly affected v1 type callback plugins that could result in a `None` on load and also cases in which extraneous python files seemed to match plugins. All this I believe we already fixed in the loader, but have not confirmed. 

Since it is preexisting code, I didn't want to remove with this change, I suspect it is not needed anymore but I would postpone removing it to a subsequent PR.

nitzmahone commented on lib/ansible/executor/task_queue_manager.py:
```diff
@@ -171,49 +171,62 @@ def load_callbacks(self):
-                    self._callback_plugins.append(callback_obj)
-                else:
-                    display.warning("Skipping '%s', unable to load or use as a callback" % callback_plugin_name)
+                    if callback_obj not in self._callback_plugins:
```
Are we doing singleton `__new__`/`__call__` magic internally on some of the callback plugin types? Seems like this would never be `False` otherwise...

bcoca replied:
'we' are not, but several 3rd party ones are, this code was introduced to account for that (mostly just moved location in this patch).

nitzmahone reviewed (commented):
Overall this looks like a much needed refactoring to better accommodate collection-hosted plugins and deal with some of the weirdness. There's a couple of things are aren't immediately obvious to me why they exist though, so if they are in fact needed, maybe some explanatory comments, and if not, well, let's get rid of them. :)

bcoca committed:
Update lib/ansible/executor/task_queue_manager.py

Co-authored-by: Matt Davis <nitzmahone@users.noreply.github.com>

bcoca committed:
added comments and warning on bad plugin

nitzmahone commented:
rebuild_merge

samdoran commented:
!waffling needs_revision

bcoca committed:
added critical whitespace

samdoran commented:
!waffling core_review
